_posts/blog/2016-09-06-model-view-viewmodel.md

```diff
@@ -0,0 +1,181 @@
+tags: [Android, Architecture, MVVM]
+date: 2016-09-04T15:39:55-04:00
+---
+After four different designs in the first six months of the development of the upday app, we learned one important lesson: we need an architecture pattern that allows fast reaction to design changes! The solution we chose in the end was Model-View-ViewModel. Discover with me what MVVM is; how we are applying it at upday and what makes it so perfect for us.
+
+## The Model-View-ViewModel Pattern
+
+The main players in the MVVM pattern are:
+
+* The *View* - that informs the ViewModel about the user's actions
+* The *ViewModel* - exposes streams of data relevant to the View
+* The *DataModel* - abstracts the data source. The ViewModel works with the DataModel to get and save the data.  
+
+At a first glance, MVVM seems very similar to the Model-View-Presenter pattern, because both of them do a great job in abstracting the view's state and behavior. The Presentation Model abstracts a view independent from a specific user-interface platform whereas the MVVM pattern was created by Microsoft to simplify the **event driven** programming of user interfaces.
+
+If, in the MVP pattern, the Presenter was telling the View directly what to display, in MVVM, **ViewModel exposes streams of events** to which the Views can bind to. Also, the Views notify the ViewModel about different actions. Thus, the MVVM pattern supports two-way data binding between the View and ViewModel and there is a many-to-one relationship between View and ViewModel. View has a reference to ViewModel but **ViewModel has no information about the View**.
+
+<center>
+<picture>
+	<a href="/images/blog/model_view_viewmodel/mvvm.png"><img src="/images/blog/model_view_viewmodel/mvvm.png" alt="Model-View-ViewModel Architecture"></a>
+	<figcaption>Model-View-ViewModel class structure</figcaption>
+</picture>
+</center>
+
+## Model-View-ViewModel at upday
+
+A quick look at the Android posts on the upday blog will instantly reveal what our favorite library is: RxJava. So it’s no wonder that RxJava is the backbone of upday's code. The event driven part required by MVVM is done using RxJava’s `Observable`s. Here's how we apply MVVM at upday, with the help of RxJava:
+
+### DataModel
+
+The DataModel exposes data easily consumable through event streams - RxJava's `Observable`s. It composes data from multiple sources, like the network layer, database or shared preferences and exposes them to whoever needs it. The DataModels hold the entire business logic.
+
+Our strong emphasis on the single responsibility principle leads to creating a DataModel for every feature in the app. For example, we have an ArticleDataModel that contains references to the API service and to the database layer. This DataModel handles the business logic ensuring that the latest news from the database is retrieved, by applying an age filter.
```

Comment:
In this bit, I'm not sure if you are referring to Java references -
> contains references

It might be better to say that -
> that composes its output from the API service and database layer.

Only a suggestion.